logrus/logrusentry.go

```diff
@@ -0,0 +1,233 @@
+	sentry "github.com/getsentry/sentry-go"
+	"github.com/sirupsen/logrus"
+)
+
+// These default log field keys are used to pass specific metadata in a way that
+// Sentry understands. If they are found in the log fields, and the value is of
+// the expected datatype, it will be converted from a generic field, into Sentry
+// metadata.
+//
+// These keys may be overridden by calling SetKey on the hook object.
+const (
+	// FieldRequest holds an *http.Request.
+	FieldRequest = "request"
+	// FieldUser holds a User or *User value.
+	FieldUser = "user"
+	// FieldTransaction holds a transaction ID as a string.
+	FieldTransaction = "transaction"
+	// FieldFingerprint holds a string slice ([]string), used to dictate the
+	// deduplication of this event.
+	FieldFingerprint = "fingerprint"
+
+	// These fields are simply omitted, as they are duplicated by the Sentry SDK.
+	FieldGoVersion = "go_version"
+	FieldMaxProcs  = "go_maxprocs"
+)
+
+// Hook is the logrus hook for Sentry.
+//
+// It is not safe to configure the hook while logging is happening. Please
+// perform all configuration before using it.
+type Hook struct {
+	client   *sentry.Client
+	scope    *sentry.Scope
+	hub      *sentry.Hub
+	fallback FallbackFunc
+	keys     map[string]string
+	levels   []logrus.Level
+}
+
+var _ logrus.Hook = &Hook{}
+
+// New initializes a new Logrus hook which sends logs to a new Sentry client
+// configured according to opts.
+func New(levels []logrus.Level, opts sentry.ClientOptions) (*Hook, error) {
+	client, err := sentry.NewClient(opts)
+	if err != nil {
+		return nil, err
+	}
+	return NewFromClient(levels, client), nil
+}
+
+// NewFromClient initializes a new Logrus hook which sends logs to the provided
+// sentry client.
+func NewFromClient(levels []logrus.Level, client *sentry.Client) *Hook {
+	h := &Hook{
+		levels: levels,
+		client: client,
+		scope:  sentry.NewScope(),
+		keys:   make(map[string]string),
+	}
+	h.setHub()
+	return h
+}
+
+// setHub refreshes the hub, to be used any time client or scope changes.
+func (h *Hook) setHub() {
+	h.hub = sentry.NewHub(h.client, h.scope)
+}
+
+// AddTags adds tags to the hook's scope.
+func (h *Hook) AddTags(tags map[string]string) {
+	h.scope.SetTags(tags)
+	h.setHub()
+}
+
+// A FallbackFunc can be used to attempt to handle any errors in logging, before
+// resorting to Logrus's standard error reporting.
+type FallbackFunc func(*logrus.Entry) error
+
+// Fallback sets a fallback function, which will be called in case logging to
+// sentry fails. In case of a logging failure in the Fire() method, the
+// fallback function is called with the original logrus entry. If the
+// fallback function returns nil, the error is considered handled. If it returns
+// an error, that error is passed along to logrus as the return value from the
+// Fire() call. If no fallback function is defined, a default error message is
+// returned to Logrus in case of failure to send to Sentry.
+func (h *Hook) Fallback(fb FallbackFunc) {
```

Comment:
In Go, this would be a getter. `Fallback/SetFallback` pairs should be used instead.

Reply:
I'm not sure I follow.  This is just a setter, not a getter.  Are you saying a getter needs to be added? I can trivally add a getter, but is there any value in such a getter?

Reply:
No no, I mean that if it is a setter (which it is), then it should be named `SetFallback`. Without `Set`, its implied getter.

Reply:
ah, got it.  I'll update.